designs/variables-for-userdata.md

```diff
@@ -0,0 +1,206 @@
+# RFC: Non-Drifting Variables for `userData` in EC2NodeClass
+
+## Overview
+
+This proposal addresses the need for users to include Kubernetes-native dynamic/short-lived data (e.g., bearer tokens) in their `EC2NodeClass`'s `userData` without triggering Karpenter's drift reconciliation mechanism. Currently, any change to the `spec.userData` field in an `EC2NodeClass` is considered drift, which flags all associated nodes for replacement. This is behaviour is undesired for AMIs using declarative, first-boot provisioning (e.g., Fedora CoreOS w/ Ignition), which expect configuration to be provided at launch time, and which cannot run arbitrary shell scripts. Additionally, this is problematic for use cases that rely on rotating credentials which **must** be stored in a Kubernetes cluster, and not in stores such as AWS Secrets Manager or SSM Parameter Store.
+
+This document proposes introducing a Kubernetes-native mechanism to template variables into the `userData`. These variables and their values would be defined in the `EC2NodeClass` but would be explicitly excluded from the drift detection logic. This allows secrets or other dynamic configuration to be updated frequently without causing a full node fleet upgrade, while still allowing intentional changes to the underlying `userData` to trigger drift as expected.
+
+## User Stories
+
+* As a cluster manager using a `Custom` AMIFamily, I want to inject short-lived secrets (like an authentication token), which are embedded in Kubernetes resources, into my node's `userData` so that nodes can be securely bootstrapped, without drift concerns.
```

Comment:
I realize theire are limitations at play here in a real system, but user-data is not secure since it's queryable via an EC2 API unencrypted. However, I could see other use-cases for templating user-data for custom AMI cases. I just wouldn't want to expose this templated var for secrets as a best practice or advisable from a security perspective.